### PR TITLE
PKey Auth Updates

### PR DESCRIPTION
**Pull request by trwalke**
Preventing MSAL from advertising PKey Auth on unsupported platforms.
Resolving OS version check bug.

Fixes for https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1849
and https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2302

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs [hunk 1]
```diff
@@ -76,8 +78,20 @@ public static bool CanOSPerformPKeyAuth()
         {
             //PKeyAuth can only be performed on operating systems with a major OS version of 6. 
             //This corresponds to windows 7, 8, 8.1 and their server equivilents.
-#if DESKTOP
-            if (Environment.OSVersion.Version.Major == 6)
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:80]: Sense Environment.OSVersion is not accurate, I used slightly different methods to properly detect the OS version. These are slightly different between .net fwk and .net core but produce the same results in testing

Comment (trwalke) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:80]: @pmaytak We used to advertise that we are PKey Auth capable on every call in the token client but now we are looking to only do this on older versions of windows. (win 8 and below).

Thus I am adding these checks here to make sure we are only adding this header on supported platforms. Do you think these checks will have any impact on performance?

File: src/client/Microsoft.Identity.Client/OAuth2/TokenClient.cs [hunk 2]
```diff
@@ -157,8 +158,11 @@ private void AddBodyParamsAndHeaders(IDictionary<string, string> additionalBodyP
                     _serviceBundle.HttpTelemetryManager.GetLastRequestHeader());
             }
 
-            //Signaling that the client can perform PKey Auth
-            _oAuth2Client.AddHeader(PKeyAuthConstants.DeviceAuthHeaderName, PKeyAuthConstants.DeviceAuthHeaderValue);
+            //Signaling that the client can perform PKey Auth on supported platforms
+            if (DeviceAuthHelper.CanOSPerformPKeyAuth())
+            {
+                _oAuth2Client.AddHeader(PKeyAuthConstants.DeviceAuthHeaderName, PKeyAuthConstants.DeviceAuthHeaderValue);
+            }
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/OAuth2/TokenClient.cs:165]: @pmaytak We used to advertise that we are PKey Auth capable on every call in the token client but now we are looking to only do this on older versions of windows. (win 8 and below).

Thus I am adding these checks here to make sure we are only adding this header on supported platforms. Do you think these checks will have any impact on performance?

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs (hunk 1 above)
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:80]: @pmaytak We used to advertise that we are PKey Auth capable on every call in the token client but now we are looking to only do this on older versions of windows. (win 8 and below).

Thus I am adding these checks here to make sure we are only adding this header on supported platforms. Do you think these checks will have any impact on performance?

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs [hunk 3]
```diff
@@ -76,8 +78,20 @@ public static bool CanOSPerformPKeyAuth()
...
+            //Environment.OSVersion as it will return incorrect information on some operating systems
+#if NET_CORE
```
Comment (jmprieur) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:82]: @trwalke 
A partner recommended in https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1824, the use of the [version helper APIs](https://docs.microsoft.com/en-us/windows/win32/sysinfo/version-helper-apis), but I suppose for PKAuth this is not needed ?

**Review by jmprieur — COMMENTED** [review]
LGTM.
I've asked a question (with a related issue)

**Comment by jennyf19** [comment]
i think we need to get the device auth tests working. [https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2207](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2207)

---
In reply to: [554694167](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2312#issuecomment-554694167) [](ancestors = 554694167)

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs (hunk 3 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:82]: correct. in the linked issue in that issue, bogdan includes a SO post w/more info.

---
In reply to: [545162569](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2312#discussion_r545162569) [](ancestors = 545162569)

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs (hunk 1 above)
Comment (pmaytak) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:80]: @trwalke Seems like `CanOSPerformPKeyAuth` ends up being called for each request, right? Also, the OS version shouldn't change. So I would probably call this method once (at app startup or with a first request) and cache the OS version in a variable somewhere in the app.

Also according to [this](https://stackoverflow.com/questions/1097363/what-are-the-performance-implications-of-accessing-a-windows-registry-value) and [this](https://devblogs.microsoft.com/oldnewthing/20060222-11/?p=32193), registry access should be fast and the recent values are cached, but there're still extra those cycles spent, compared to if we cache the value in the app.

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs (hunk 3 above)
Comment (pmaytak) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:82]: I would probably just use the registry once and cache the OS version in the app. Seems like a simpler approach.

Interestingly the [version helper APIs](https://docs.microsoft.com/en-us/windows/win32/sysinfo/version-helper-apis) are recommened for this. But then they use VerifyVersionInfo function which is [deprecated](https://docs.microsoft.com/en-us/windows/win32/sysinfo/targeting-your-application-at-windows-8-1) in Windows 10 but I guess, can still work if the app has app.manifest.

**Review by pmaytak — APPROVED** [review]

**Comment by trwalke** [comment]
Currently waiting for customer to test the preview package for this fix

File: src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj [hunk 4]
```diff
@@ -257,4 +257,8 @@
     <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0" PrivateAssets="All" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Reference Include="System.Management" />
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:261]: is this for the registry stuff? if so, please add it to the desktop-only ItemGroup

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs [hunk 5]
```diff
@@ -7,14 +7,18 @@
...
 using Microsoft.Identity.Client.Http;
 using Microsoft.Identity.Client.PlatformsCommon.Shared;
 using Microsoft.Identity.Client.Utils;
+using Microsoft.Win32;
 
 namespace Microsoft.Identity.Client.PlatformsCommon
 {
     internal class DeviceAuthHelper
     {
+        private static bool? s_canOSPerformPKeyAuth;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:20]: maybe use the Lazy<T> initalization as it's thread safe (although this is probably ok as well)

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs [hunk 6]
```diff
@@ -74,14 +78,34 @@ public static string GetBypassChallengeResponse(Dictionary<string, string> respo
 
         public static bool CanOSPerformPKeyAuth()
         {
+            if (s_canOSPerformPKeyAuth != null)
+            {
+                return (bool)s_canOSPerformPKeyAuth;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:83]: s_canOSPerformPKeyAuth.Value

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs (hunk 1 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:80]: Please add the link where you copied the registry solution to the code, it's helpful to have a reference

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs [hunk 7]
```diff
@@ -74,14 +78,34 @@ public static string GetBypassChallengeResponse(Dictionary<string, string> respo
...
+            }
+
             //PKeyAuth can only be performed on operating systems with a major OS version of 6. 
             //This corresponds to windows 7, 8, 8.1 and their server equivilents.
-#if DESKTOP
-            if (Environment.OSVersion.Version.Major == 6)
+            //Environment.OSVersion as it will return incorrect information on some operating systems
+#if NET_CORE
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:89]: #if NET_CORE || NET5_WIN

File: src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs [hunk 8]
```diff
@@ -74,14 +78,34 @@ public static string GetBypassChallengeResponse(Dictionary<string, string> respo
...
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                && !RuntimeInformation.OSDescription.Contains("Windows 10"))
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/DeviceAuthHelper.cs:91]: please use a case-insensitive Contains

**Review by bgavrilMS — COMMENTED** [review]
LGTM with some small changes, particularly the new reference

**Comment by trwalke** [comment]
Currently waiting for customer to test the preview package for this fix. The customer says that they are encountering build issues when trying to test fix.

However, I am able to test this locally and it works.

**Comment by henrik-me** [comment]
>     It should not be required after the december release of .NET 5 SDK -->

@bogdan, do you have a work item tracking this?

---
Refers to: src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:160 in 6344d7c. [](commit_id = 6344d7cbe502e289f74003fddca95bc9533ce344, deletion_comment = False)

**Comment by bgavrilMS** [comment]
@trwalke - can we merge this? It's been sitting here for 28 days. I think this is a reasonable fix and @pmaytak reviewed it.

**Review by bgavrilMS — APPROVED** [review]

**Pull request merged**